rivertype/river_type.go

```diff
@@ -229,6 +230,49 @@ type AttemptError struct {
 	Trace string `json:"trace"`
 }
 
+type JobInsertParams struct {
+	Args         JobArgs
+	CreatedAt    *time.Time
+	EncodedArgs  []byte
+	Kind         string
+	MaxAttempts  int
+	Metadata     []byte
+	Priority     int
+	Queue        string
+	ScheduledAt  *time.Time
+	State        JobState
+	Tags         []string
+	UniqueKey    []byte
+	UniqueStates byte
+}
+
+// JobInsertMiddleware provides an interface for middleware that integrations can
+// use to encapsulate common logic around job insertion.
+//
+// Implementations should embed river.JobMiddlewareDefaults to inherit default
+// implementations for phases where no custom code is needed, and for forward
+// compatibility in case new functions are added to this interface.
+type JobInsertMiddleware interface {
+	// InsertMany is invoked around a batch insert operation. Implementations
+	// must always include a call to doInner to call down the middleware stack
+	// and perfom the batch insertion, and may run custom code before and after.
+	//
+	// Returning an error from this function will fail the overarching insert
+	// operation, even if the inner insertion originally succeeded.
+	InsertMany(ctx context.Context, manyParams []*JobInsertParams, doInner func(context.Context) ([]*JobInsertResult, error)) ([]*JobInsertResult, error)
```

Comment:
@brandur I don't think we ever discussed why these middleware functions don't pass through the `manyParams` or `job` as arguments to `doInner`. I'm guessing it's partly to do with making it easier & cleaner for callers?

I don't think there's a use case for changing out the job struct on the `Work` variant, but for insertion it's conceivable that one could want to add or remove rows to the insert batch as part of a middleware call (in addition to merely manipulating them before insertion).

Something to consider adjusting as this bakes in an RC release.